File: core/esmf-aspect-meta-model-java/src/main/java/org/eclipse/esmf/metamodel/characteristic/impl/DefaultEnumeration.java
/*
 * Copyright (c) 2024 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */
package org.eclipse.esmf.metamodel.characteristic.impl;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.StringJoiner;

import org.eclipse.esmf.metamodel.characteristic.Enumeration;
import org.eclipse.esmf.metamodel.Type;
import org.eclipse.esmf.metamodel.Value;
import org.eclipse.esmf.metamodel.impl.DefaultCharacteristic;
import org.eclipse.esmf.aspectmodel.loader.MetaModelBaseAttributes;
import org.eclipse.esmf.aspectmodel.visitor.AspectVisitor;

@SuppressWarnings( "squid:S1150" ) // Sonar thinks this implements java.util.Enumeration, which it does not
public class DefaultEnumeration extends DefaultCharacteristic implements Enumeration {
   private final List<Value> values;

   public DefaultEnumeration( final MetaModelBaseAttributes metaModelBaseAttributes, final Type dataType, final List<Value> values ) {
      super( metaModelBaseAttributes, Optional.of( dataType ) );
      this.values = values;
   }

   /**
    * A list of valid states.
    *
    * @return the values.
    */
   @Override
   public List<Value> getValues() {
      return values;
   }

   /**
    * Accepts an Aspect visitor
    *
    * @param visitor The visitor to accept
    * @param <T> The result type of the traversal operation
    * @param <C> The context of the visitor traversal
    */
   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitEnumeration( this, context );
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultEnumeration.class.getSimpleName() + "[", "]" )
            .add( "values=" + values )
            .toString();
   }

   @Override
   public boolean equals( final Object o ) {
      if ( this == o ) {
         return true;
      }
      if ( o == null || getClass() != o.getClass() ) {
         return false;
      }
      if ( !super.equals( o ) ) {
         return false;
      }
      final DefaultEnumeration that = (DefaultEnumeration) o;
      return Objects.equals( values, that.values );
   }

   @Override
   public int hashCode() {
      return Objects.hash( super.hashCode(), values );
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/esmf/aspect-meta-model/characteristic/default-enumeration.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */
import {Type} from '../type';
import {Characteristic, DefaultCharacteristic} from './default-characteristic';
import {DefaultEntityInstance} from '../default-entity-instance';
import {DefaultEntity} from '../default-entity';
import { DefaultCollectionValue } from '../default-collection-value';
import { DefaultScalarValue } from '../default-scalar-value';

export interface Enumeration extends Characteristic {
    values: Array<DefaultEntityInstance | DefaultCollectionValue | DefaultScalarValue| string | number>;

    indexOf(value: string): number;
}

export class DefaultEnumeration extends DefaultCharacteristic implements Enumeration {
    constructor(
        metaModelVersion: string,
        aspectModelUrn: string,
        name: string,
        private _values: Array<DefaultEntityInstance | DefaultCollectionValue | DefaultScalarValue| string | number>,
        dataType?: Type
    ) {
        super(metaModelVersion, aspectModelUrn, name, dataType);
    }

    public set values(values: Array<DefaultEntityInstance | DefaultCollectionValue | DefaultScalarValue| string | number>) {
        this._values = values;
    }

    public get values(): Array<DefaultEntityInstance | DefaultCollectionValue | DefaultScalarValue| string | number> {
        return this._values;
    }

    /**
     * Find the index of the given value in the values array.
     *
     * @return returns the index of the value or -1
     */
    public indexOf(value: string): number {
        if (!this.values) {
            return -1;
        }

        if (this.dataType && this.dataType.isComplex) {
            const propertyValue = (<DefaultEntity>this.dataType).properties.find(property => property.isNotInPayload === false);

            return this.values.findIndex((valueEntry: DefaultEntityInstance) => {
                return valueEntry[propertyValue.name] === value;
            });
        } else {
            return this.values.findIndex(valueEntry => valueEntry === value);
        }
    }
}
